Improvements to commands.config

alexfikl opened the pull request:
Was trying to figure out what configs were used in my venv and fell down a little rabbit hole!

This make some updates to the ``config`` command:
* Improve the docs a bit.
* Do not show scary tracebacks on errors, but some hopefully helpful error messages.
* Allow more positional arguments to query multiple settings at once.
* Added `--default` to show all the defaults papis knows about for a given section (could be helpful to explore the settings without going into the code)
* Added a `--json` flag to print multiple options in the json format instead of the default INI-ish format.

@jghauser if you have the time to take a look at the documentation parts (phrasing and such), it would be greatly appreciated, as usual!

jghauser commented:
This is great, love the ability to query multiple settings at once. I use this in papis.nvim and hopefully it will speed up things a bit! I'll have a look at the docs!

alexfikl commented:
> This is great, love the ability to query multiple settings at once. I use this in papis.nvim and hopefully it will speed up things a bit! I'll have a look at the docs!

@jghauser Ah! This changed the output to show `<key> = <value>` instead of just `<value>` for each option. Will that break stuff for you? 

What would you be a useful format to list more options? Add a `--json` flag or just list them in order?

jghauser commented:
Not a big deal if things break here, I can easy fix things! Though maybe nicer if the output remains the same when querying single settings. I only use this when the user hasn't copied over the settings explicitly to neovim (which I recommend people do as querying papis is rather slow and slows down neovim startup).

alexfikl commented:
> Not a big deal if things break here, I can easy fix things! Though maybe nicer if the output remains the same when querying single settings. I only use this when the user hasn't copied over the settings explicitly to neovim (which I recommend people do as querying papis is rather slow and slows down neovim startup).

The output for a single option should be restored now!

If you want a specific format for showing multiple options, let me know! Although that can probably be extended in the future with some flags, so not a particularly urgent issue.

jghauser commented:
Yeah, I think for now this is good. I do have some ideas for betting settings management in papis.nvim and might get back to this when I get around it (I think json might come in handy!).

alejandrogallo commented:
this is great !

jghauser commented:
What's the best way for me to submit changes to the docs here @alexfikl ? should I just edit the relevant files and commit directly to this fork (that seems to be possible)? I feel like adding suggestions is a bit too laborious an approach for prose...

jghauser commented:
Oh oh, missed a trailing whitespace. Lemme fix this quickly!

jghauser commented:
Ok, hopefully tests will pass now now. While updating the docs, two things confused me a little.

The `config` command usually queries setting's values, but with the `--list-defaults` argument, it's a section's values. Would it maybe make more sense to allow querying both sections and individual settings in some consistent way?

Also, when querying a setting that is set to an empty string, it might look to a user as if nothing was returned. I'm thinking maybe your earlier choice to return "key = value" pairs is better after all.

alexfikl commented:
Thanks a lot for the help with the docs! It looks a lot clearer and easier to follow now :heart: 

> The `config` command usually queries setting's values, but with the `--list-defaults` argument, it's a section's values. Would it maybe make more sense to allow querying both sections and individual settings in some consistent way?

Yeah, that cross my mind too, but I wasn't sure how to change it. Some things that come to mind
1. When calling with `--list-defaults`, just print the defaults in all the sections.
2. Add a flag to explicitly mention the section, which could also benefit querying more options in the same section, e.g.
```
papis config --section <section> <key>
``` 
as an alternative to
```
papis config <section>.<key>
```
? And then we can call
```
papis config --section <section> --list-defaults
```

That looks a bit more consistent to me, but suggestions are very welcome! I'm inclined to choose the first for simplicity. Would you have a use case for the more granular queries?

> Also, when querying a setting that is set to an empty string, it might look to a user as if nothing was returned. I'm thinking maybe your earlier choice to return "key = value" pairs is better after all.

Hm, we could add a warning? I'll have to check if logger warnings go to `stderr`.
Otherwise, I'm not too bothered with just leaving it be, since most cli tools tend to go by "no news is good news!".

jghauser commented:
I like your second proposal, it's very consistent. Would that then also allow querying individual defaults? As in, would `papis config <key> --list-defaults` be admissible? Also, now that I'm thinking about it, it would also be neat to have `papis config` (without arguments) spit out a list of all (and with `--list-defaults` all default) settings.

And you're right about the "no news is good news!" thing. It's how most CLI tools do it, so we should probably follow that.

alexfikl commented:
> I like your second proposal, it's very consistent. Would that then also allow querying individual defaults? As in, would `papis config <key> --list-defaults` be admissible? Also, now that I'm thinking about it, it would also be neat to have `papis config` (without arguments) spit out a list of all (and with `--list-defaults` all default) settings.

I like that as a workflow! I'll get on it :D

alexfikl commented:
Ok, I reworked it a bit to be more in line with what you proposed. It should be fairly complete now!

To summarize a bit:
* Querying specific settings still works with `papis config some-option` and adding `--default` to that will instead print the default value.
* Querying sections also works with `papis config -s some-section`, which prints all the values in that section + adding `--default` prints all the defaults instead + adding options prints specific options from the section.
* Querying multiple values + multiple defaults also works just the same.
* Just calling `papis config` will print all the settings in all the sections + with support for `--default` as well.

@jghauser I updated the docs to describe these new options, so another look is probably warranted :D

jghauser commented:
This is great! I've been playing around with it for a bit and have some comments/questions:

- What is shown when querying without `--default`? I feel like as a user my assumption would be that it shows whatever Papis is using, so either the explicitly set setting or, if that doesn't exist, the default (whereas with `--default`, always the default is shown). But I think what's happening is that a setting is only shown if it's explicitly set? But I might be wrong with my expectations here, what do you think? In either case, I'll make sure to document it.
- I think the settings I set in my user `config.py` are counted as default. Is that necessitated by how settings are set by Papis or could we change that? If it's not (easily) changeable, I'll make a note of it in the docs.

alexfikl commented:
>     * What is shown when querying without `--default`? I feel like as a user my assumption would be that it shows whatever Papis is using, so either the explicitly set setting or, if that doesn't exist, the default (whereas with `--default`, always the default is shown). But I think what's happening is that a setting is only shown if it's explicitly set? But I might be wrong with my expectations here, what do you think? In either case, I'll make sure to document it.

If querying without `--section`, it should work the way you described, i.e.
```
papis config <key> 				-> either config file value or default
papis config --default <key>	-> always default
```

However, when querying with a section, it just shows stuff from the configuration file, i.e.
```
papis config --section <sec>		  	-> all keys in the configuration file (no defaults)
papis config --default --section <sec>	-> all defaults (whether in the config file or not)
``` 

If you add a key and a section, i.e. `papis config --section <sec> <key>`, it should still pick up the default if there is one. Is that what you're seeing as well?

>     * I think the settings I set in my user `config.py` are counted as default. Is that necessitated by how settings are set by Papis or could we change that? If it's not (easily) changeable, I'll make a note of it in the docs.

Hm, I only use the static INI file for the configuration, so haven't had any interaction with the `config.py` thing. I'll take a look if that does something different and get back to you!

jghauser commented:
Ok, thanks for the explanation!

> If you add a key and a section, i.e. papis config --section <sec> <key>, it should still pick up the default if there is one. Is that what you're seeing as well?

Yes, I get the same behaviour!

I think it would be better if adding --section didn't change the behaviour re showing only explicitly set settings without `--defaults`.  To keep behaviour consistent across versions of the command, I would prefer the command without `--defaults` to always show whatever settings Papis is using (explicitly set or default). Or maybe I'm missing something -- do you prefer it the way it's set up now? Though I don't think this is super crucial and if tedious to implement, I'll simply make a note in the docs :).

alexfikl commented:
> I would prefer the command without --defaults to always show whatever settings Papis is using (explicitly set or default).

Agreed that that's what we want to go for! When is it not doing that? Can you given an example?

jghauser commented:
> Agreed that that's what we want to go for! When is it not doing that? Can you given an example?

What I mean is the difference between these two:
- `papis config <key>`
- `papis config --section <sec>`

Where the first gives either explicitly set settings or defaults and the second only gives the explicitly set stuff. I would intuitively interpret the second command as implying the not returned settings aren't set at all (or to nil). In that instance, if I understand things correctly, `papis config` would not give me the values used by Papis.

> EDIT: Also, I looked a bit at using a config.py file and it seems like you can just put anything in there and it'll get exec-ed. I don't think we can do much about that if users modify the default settings in there..

Fair enough, that makes sense. Given that the `config.py` is an advanced thing anyway, I don't think this is an issue. I'll make a note about it in the docs.

alexfikl commented:
> What I mean is the difference between these two:
> 
>     `papis config <key>`
>     `papis config --section <sec>`
> 
> 
> Where the first gives either explicitly set settings or defaults and the second only gives the explicitly set stuff. I would intuitively interpret the second command as implying the not returned settings aren't set at all (or to nil). In that instance, if I understand things correctly, `papis config` would not give me the values used by Papis.

Ah, I see! Thank you for explaining and sorry for being a bit dense about it!

That shouldn't be too hard to implement :rocket:

jghauser commented:
Nothing to be sorry about, you're doing all the hard work here :)

alexfikl commented:
Ok, I think this does what you suggested now! For an example, using the [bibtex section settings](https://github.com/papis/papis/blob/287f2d8199e6ad85e4a5c31cfde1e809a15dd7a1/papis/commands/bibtex.py#L152-L156), I get
```
> papis config --default --section bibtex
[bibtex]
default-read-bibfile = ''
auto-read = ''
default-save-bibfile = ''
```
and without the defaults
```
> papis config --section bibtex
[bibtex]
default-read-bibfile = 'references.bib'
auto-read = 'true'
default-save-bibfile = ''
```
which is updated with the settings I have in my configuration file. Does it work as expected now?

jghauser commented:
:tada: I checked it out and it's now really working exactly like I suggested. Thanks! :)

alexfikl commented:
> I checked it out and it's now really working exactly like I suggested. Thanks! :)

Awesome! Thanks a lot for all the help with this!

jghauser commented on papis/commands/config.py:
```diff
@@ -1,34 +1,60 @@
+With ``-l``, Papis selects a specific library (here, the "books" library). The
+rest works just as above.
+
+Settings from a specific section in the configuration file can also be
+accessed. To take an example, the `Bibtex`_ command's settings can be
+accessed with
```
```suggestion
You can also query a specific setting within a section. For example like this:
```

jghauser commented on papis/commands/config.py:
```diff
@@ -1,34 +1,60 @@
+
+.. code::
+
+    papis config --default --section <name>
+    papis config --section <name>
```
```suggestion
    papis config --section bibtex
    papis config --default --section bibtex
```

jghauser commented on papis/commands/config.py:
```diff
@@ -1,34 +1,60 @@
 
-    papis config match-format
-    > {doc[tags]}{doc.subfolder}{doc[title]}{doc[author]}{doc[year]}
+    papis config --default --section bibtex default-read-bibfile
+    papis config --section bibtex default-read-bibfile
```
```suggestion
    papis config --section bibtex default-read-bibfile
    papis config --default --section bibtex default-read-bibfile
```

jghauser commented on papis/commands/config.py:
```diff
@@ -1,34 +1,60 @@
+    papis config --section bibtex default-read-bibfile
 
-You can find a list of all available settings in the configuration section.
+For some more advanced usage, we can also query multiple settings at once. Then
+it may be useful to overwrite the selected section for several of the settings.
+This can be achieved by
```
```suggestion
Alternatively, you can also you the ``<section.setting>`` format to query the value of
a setting in a specific `section`:

.. code::
    papis config bibtex.default-read-bibfile
    papis config --default bibtex.default-read-bibfile
    
For some more advanced usage, we can also query multiple settings at once.  Here,
sections specified with ``<section>.<setting>`` override the section specified by
``--setting <section>``. This can be achieved by:
```

jghauser reviewed (commented):
Finally had time to go through the docs and made some changes. I've been trying out the comments functionality, but it's really a bit clunky for more than typo fix in the case of prose. It's really quite hard to keep an overview of the whole doc. Hope I didn't mess up anything :sweat_smile:

alexfikl commented on papis/commands/config.py:
```diff
@@ -1,34 +1,60 @@
-it with
+With ``-l``, Papis selects a specific library (here, the "books" library). The
+rest works just as above.
+
```
Yep, you're right. Removed it in [8c81871](https://github.com/papis/papis/pull/454/commits/8c81871e5c115e5c79cdc79d410b8e34f77e2495).

alexfikl commented:
This looks good to me now, so I went ahead and merged it! Thanks a lot for all the help and suggestions :heart:

alejandrogallo commented:
wow this is really fantastic, great job you two!!
